### ISSUE TITLE
Paddleseg framework compatibility issue and some possible fix

### ISSUE THREAD

**Issue by Leev1s** [rankseg/rankseg#15]
## Paddleseg

The original paddleseg predict function use `pred, _ = infer()`. This is done to save memory. The various application examples provided by the Paddleseg project are more geared towards industrial use cases.

- [PP-HumanSegV2 Official Demo](https://aistudio.baidu.com/aistudio/projectdetail/4504982?contributionType=1)
- [PP-LiteSeg Official Demo](https://aistudio.baidu.com/aistudio/projectdetail/3873145?contributionType=1)

**Use Example**

```shell
!python ../../tools/predict.py \
--config configs/human_pp_humansegv2_lite.yml \
--model_path pretrained_models/human_pp_humansegv2_lite_192x192_pretrained/model.pdparams \
--image_path data/images/human.jpg \
--save_dir ./data/images_result
```

[tools/predict.py](https://github.com/PaddlePaddle/PaddleSeg/blob/release/2.10/tools/predict.py)

This function is most used tool func to take arg inputs in terminal.🤯 It use predetermined config file to avoid complex pipeline construction.🚧 It imports core.predict to infer.

```python
def main(args):
    assert args.config is not None, \
        'No configuration file specified, please set --config'
    cfg = Config(args.config)
    builder = SegBuilder(cfg)
    test_config = merge_test_config(cfg, args)

    utils.show_env_info()
    utils.show_cfg_info(cfg)
    if args.device != 'cpu':
        device = f"{args.device}:{args.device_id}"
    else:
        device = args.device
    utils.set_device(device)

    model = builder.model
    transforms = Compose(builder.val_transforms)
    image_list, image_dir = get_image_list(args.image_path)
    logger.info('The number of images: {}'.format(len(image_list)))

    predict(
        model,
        model_path=args.model_path,
        transforms=transforms,
        image_list=image_list,
        image_dir=image_dir,
        save_dir=args.save_dir,
        **test_config)
```

[core/predict.py](https://github.com/PaddlePaddle/PaddleSeg/blob/release/2.10/paddleseg/core/predict.py)

This is where the pred generated.🥸

```python
with paddle.no_grad():
        for i, im_path in enumerate(img_lists[local_rank]):
            data = preprocess(im_path, transforms)

            if aug_pred:
                pred, _ = infer.aug_inference(
                    model,
                    data['img'],
                    trans_info=data['trans_info'],
                    scales=scales,
                    flip_horizontal=flip_horizontal,
                    flip_vertical=flip_vertical,
                    is_slide=is_slide,
                    stride=stride,
                    crop_size=crop_size,
                    use_multilabel=use_multilabel)
            else:
                pred, _ = infer.inference(
                    model,
                    data['img'],
                    trans_info=data['trans_info'],
                    is_slide=is_slide,
                    stride=stride,
                    crop_size=crop_size,
                    use_multilabel=use_multilabel)
            pred = paddle.squeeze(pred)
            pred = pred.numpy().astype('uint8')
```

## One possible but not elegant solution

We just take what we want. Also import torch.😘

[modified/core/predict.py](https://github.com/Leev1s/rankseg/blob/main/rankseg/compatibility/paddleseg/core/predict.py)
```python
with paddle.no_grad():
        for i, im_path in enumerate(img_lists[local_rank]):
            data = preprocess(im_path, transforms)

            if aug_pred:
                pred, logit = infer.aug_inference(
                    model,
                    data['img'],
                    trans_info=data['trans_info'],
                    scales=scales,
                    flip_horizontal=flip_horizontal,
                    flip_vertical=flip_vertical,
                    is_slide=is_slide,
                    stride=stride,
                    crop_size=crop_size,
                    use_multilabel=use_multilabel)
            else:
                pred, logit = infer.inference(
                    model,
                    data['img'],
                    trans_info=data['trans_info'],
                    is_slide=is_slide,
                    stride=stride,
                    crop_size=crop_size,
                    use_multilabel=use_multilabel)

            if use_rankseg:
                import torch
                from rankseg import RankSEG
                
                # Paddle logit → PyTorch probs
                logit_np = logit.numpy()
                logit_torch = torch.from_numpy(logit_np)
                probs = torch.softmax(logit_torch, dim=1)
                
                # RankSEG optimization
                rankseg = RankSEG(
                    metric=rankseg_metric,
                    solver=rankseg_solver,
                    output_mode=rankseg_output_mode
                )
                pred = rankseg.predict(probs)
                
                # Convert back to numpy
                if rankseg_output_mode == 'multiclass':
                    pred = pred.squeeze().numpy().astype('uint8')
                else:
                    # For multilabel, keep the channel dimension for visualization
                    # Assuming pred is (B, C, H, W) and B=1
                    pred = pred.squeeze(0).numpy().astype('uint8')
            else:
                pred = paddle.squeeze(pred)
                pred = pred.numpy().astype('uint8')
```

[modified/tools/predict](https://github.com/Leev1s/rankseg/blob/main/rankseg/compatibility/paddleseg/tools/predict.py)

Add some var to accept additional rankseg args. And add some other necessary components... It just work. I only test it on single pic single class infer. From the output pic, rankseg really improved the result.

```python
def merge_test_config(cfg, args):
    test_config = cfg.test_config
    if 'aug_eval' in test_config:
        test_config.pop('aug_eval')
    if 'auc_roc' in test_config:
        test_config.pop('auc_roc')
    if args.aug_pred:
        test_config['aug_pred'] = args.aug_pred
        test_config['scales'] = args.scales
        test_config['flip_horizontal'] = args.flip_horizontal
        test_config['flip_vertical'] = args.flip_vertical
    if args.is_slide:
        test_config['is_slide'] = args.is_slide
        test_config['crop_size'] = args.crop_size
        test_config['stride'] = args.stride
    if args.custom_color:
        test_config['custom_color'] = args.custom_color
    if args.use_multilabel:
        test_config['use_multilabel'] = args.use_multilabel
    # RankSEG parameters
    if args.use_rankseg:
        test_config['use_rankseg'] = args.use_rankseg
        test_config['rankseg_metric'] = args.rankseg_metric
        test_config['rankseg_solver'] = args.rankseg_solver
        test_config['rankseg_output_mode'] = args.rankseg_output_mode
    return test_config
```

|argmax|rankseg|
|--|--|
| <img src="https://s2.loli.net/2026/02/06/ND9P4vUMFyHWJag.jpg"> | <img src="https://s2.loli.net/2026/02/06/XpgDSInbamuhc51.jpg"> |

Now this solution is put in rankseg.compatibility in my repo.

`from rankseg.compatibility.paddleseg.core.predict import predict`

Things should think about:
- Can rankseg accept paddle tensor?
- Should add compatibility to none config file usage? The traditional `with paddle.no_grad():` case.
- The paddle demo for rankseg repo should use command usage example or pure python code? I just don't know how others using paddle. Need more research about it.

## To be continue.

**Comment by statmlben** [comment]
@Leev1s Thank you for the great contribution! From my perspective, the current implementation can interface with Paddle, but it is not an elegant solution. To improve this, we need paddle to expose the `logit` outputs, allowing us to integrate RankSEG through a more systematic post-processing approach.

Let's revise our strategy: On our side, we can first implement paddle's framework (as you have done) and provide compelling evidence (benchmarks or examples) to raise an issue with the paddle team. We will demonstrate that RankSEG can enhance performance and request that they expose `logit` outputs to enable more systematic post-processing. Of course, we are very willing to collaborate with them to develop this functionality.

**Comment by Leev1s** [comment]
Thanks for the direction @statmlben! I want to correct something from my original post — the `pred, _ = infer()` issue actually has a cleaner solution than I first thought.

## A more appropriate approach: `_generate_pred_from_logit`

The key insight is that we don't need to change the `infer()` signature at all. Instead, we can introduce a small post-processing helper inside `infer.py` that sits between the logit and the argmax step:

```python
def _generate_pred_from_logit(logit,
                               use_multilabel=False,
                               use_rankseg=False,
                               rankseg_metric='dice',
                               rankseg_solver='RMA',
                               rankseg_output_mode='multiclass',
                               **rankseg_kwargs):
    if not use_rankseg:
        # Original PaddleSeg logic — untouched
        if not use_multilabel:
            pred = paddle.argmax(logit, axis=1, keepdim=True, dtype='int32')
        else:
            pred = (F.sigmoid(logit) > 0.5).astype('int32')
    else:
        # RankSEG: logit (Paddle) -> probs (PyTorch) -> optimised pred
        import torch
        from rankseg import RankSEG

        probs = torch.softmax(torch.from_numpy(logit.numpy()), dim=1)
        rankseg = RankSEG(metric=rankseg_metric, solver=rankseg_solver,
                          output_mode=rankseg_output_mode, **rankseg_kwargs)
        pred_np = rankseg.predict(probs).cpu().numpy()

        if rankseg_output_mode == 'multiclass':
            if pred_np.ndim == 3:
                pred_np = pred_np[:, np.newaxis, :, :]
        pred = paddle.to_tensor(pred_np, dtype='int32')

    return pred
```

Both `inference()` and `aug_inference()` in `infer.py` call `_generate_pred_from_logit` at the end, passing through the RankSEG flags. The existing `pred, _ = infer()` call signature is **preserved** — callers don't need to change anything. The logit is still computed internally, RankSEG just intercepts it before the final argmax.

The full implementation is at [`rankseg/paddleseg/infer.py`](https://github.com/Leev1s/rankseg/blob/paddleseg/rankseg/paddleseg/infer.py).

## Benchmark (CamVid, PP-LiteSeg-T, single-scale)

| Model | mIoU | Dice |
|---|---|---|
| PP-LiteSeg-T (baseline) | 75.92% | 81.38% |
| PP-LiteSeg-T + RankSeg | **76.13%** | **82.48%** |

+0.21 pp mIoU and +1.10 pp Dice with zero retraining, reproduced with a pre-built Docker image — see the [README](https://github.com/Leev1s/rankseg/blob/paddleseg/rankseg/paddleseg/README.md) for exact commands.

**Comment by ZixunWang** [comment]
@Leev1s Thanks for the great work! 

Just curious: whether the IoU/Dice computed in Paddleseg is accumulated over the whole dataset or just one image. Please refer to the $IoU^{I}$ vs. $IoU^{D}$ in RankSEG-RMA paper.

**Comment by Leev1s** [comment]
@ZixunWang Thanks for the great question!

After carefully analyzing PaddleSeg validation logic in `val.py`, I can confirm that **PaddleSeg computes IoU/Dice using dataset-level accumulation ($IoU^D$), not image-level averaging ($IoU^I$)**.

## How PaddleSeg Calculates IoU

The implementation in `val.py` follows this approach:

1. **Initialize accumulation variables** for the entire dataset:
   ```python
   intersect_area_all = paddle.zeros([1], dtype='int64')
   pred_area_all = paddle.zeros([1], dtype='int64')
   label_area_all = paddle.zeros([1], dtype='int64')
   ```

2. **Accumulate statistics from all images** in the loop:
   ```python
   for i in range(len(intersect_area_list)):
       intersect_area_all = intersect_area_all + intersect_area_list[i]
       pred_area_all = pred_area_all + pred_area_list[i]
       label_area_all = label_area_all + label_area_list[i]
   ```

3. **Calculate metrics using accumulated statistics**:
   ```python
   metrics_input = (intersect_area_all, pred_area_all, label_area_all)
   class_iou, miou = metrics.mean_iou(*metrics_input)
   class_dice, mdice = metrics.dice(*metrics_input)
   ```

This corresponds to **$IoU^D$** (dataset-level) metric in the RankSEG-RMA paper, where:
- $IoU^D = \frac{\sum_{i=1}^{N} TP_i}{\sum_{i=1}^{N} (TP_i + FP_i + FN_i)}$

## Comparison with Image-Level IoU ($IoU^I$)

For completeness, I've just added image-level IoU calculation to `val.py` so we can compare both metrics:

```python
# Calculate image-level IoU (IoU^I)
class_iou_per_img, miou_per_img = metrics.mean_iou(
    intersect_area, pred_area, label_area)
miou_per_image_list.append(miou_per_img)

# At the end:
miou_image_level = np.mean(miou_per_image_list)
```

This corresponds to **$IoU^I$** (image-level) metric:
- $IoU^I = \frac{1}{N} \sum_{i=1}^{N} \frac{TP_i}{TP_i + FP_i + FN_i}$

## Key Differences

- **$IoU^D$ (PaddleSeg default)**: Weighted by pixel count
- **$IoU^I$**: Average across images, treats all images equally regardless of size

The benchmark results I reported (mIoU: 75.92% → 76.13%, Dice: 81.38% → 82.48%) were computed using **$IoU^D$**, which is standard in PaddleSeg.

```
[2026/03/03 14:19:49] INFO: RankSEG enabled: metric=iou, solver=RMA, output_mode=multiclass
[2026/03/03 14:19:49] INFO: Start evaluating (total_samples: 101, total_iters: 101)...
101/101 [==============================] - 70s 690ms/step - batch_cost: 0.6898 - reader cost: 5.7615e-04
[2026/03/03 14:20:58] INFO: [EVAL] #Images: 101 mIoU: 0.7622 Acc: 0.9604 Kappa: 0.9506 Dice: 0.8255
[2026/03/03 14:20:58] INFO: [EVAL] Image-level average mIoU (IoU^I): 0.7377
```
```
[2026/03/03 14:15:18] INFO: Start evaluating (total_samples: 101, total_iters: 101)...
101/101 [==============================] - 15s 147ms/step - batch_cost: 0.1473 - reader cost: 5.8761e-04
[2026/03/03 14:15:33] INFO: [EVAL] #Images: 101 mIoU: 0.7592 Acc: 0.9652 Kappa: 0.9564 Dice: 0.8138
[2026/03/03 14:15:33] INFO: [EVAL] Image-level average mIoU (IoU^I): 0.7361
```

**Comment by ZixunWang** [comment]
@Leev1s Thank you for the clear clarification!

**Comment by statmlben** [comment]
Thank you @Leev1s for the careful review and detailed validation evaluation!
